### DISCUSSION TITLE
Extending "users" Collection via External File upload (.CSV or .JSON) ?

### DISCUSSION THREAD

**Discussion by MEF-01** [pocketbase/pocketbase#2105] [Q&A]
We want to pre-register approx 200 new "users" for a test React App using PB as the back-end. For each new user, we want register their Email + (temporary)  Password  - this data is held in a 2 column .CSV file on desktop.
Q. Is it possible to upload a file containing the 200 new "users" (ideally in .CSV format,  or .CSV converted to .JSON format) 
direct to the existing "users" Collection to extend it ? ie without to avoid having to manually create 200 new records ?  (NB at a later stage users will DIY Sign-up via the App front-end, but this not existing yet at the testing stage) Any pointers gratefully received.

**Comment by ganigeorgiev** [comment]
In order to be able to import users from CSV/JSON, each user entry must have minumim:

- `id` - random string, preferrable a 15 chars string, but less works too (in the next release this will be auto generated on db level if missing)
- `created` - datetime string (in the next release this will be auto generated on db level if missing)
- `updated` - datetime string (in the next release this will be auto generated on db level if missing)
- `username` - random unique string
- `tokenKey` - must be cryptographically random string of 30+ chars
- `passwordHash` - must be a bcrypt hash string (12 round/cost factor would be fine)

Once you've prepared the CSV/JSON, you can import it directly into the `pb_data/data.db` db file using the [sqlite3 cli tool](https://sqlite.org/cli.html#importing_files_as_csv_or_other_formats), something like:
```sh
sqlite3 pb_data/data.db '.import users.csv users --csv'
```

Alternatively, you can also insert the JSON items via a node JS script using the JS SDK. For an example, you can check https://github.com/pocketbase/pocketbase/discussions/2088.


---

_Side note: if you decide to import the users via the web APIs please keep in mind that we currently use 13 as bcrypt round factor which could be a little too much for low-spec hardware and could be slow (~3-4 hashes per second on 2vCPU) and in the next release it will be lowered to 12._

**Reply by nuts-n-bits** [reply]
Unrelated to original post, but can't resist asking why not bcrypt on the client and only salt on the server?

**Reply by ganigeorgiev** [reply]
@nuts-n-bits What's the benefit? You still need to store the hash in the database and verify it on login against the plain password.

Additionally, most bcrypt implementations don't use a user provided salt since it is generated and stored together with the cipher text as one string.

**Reply by nuts-n-bits** [reply]
emmm, so that you don't need server resource to run a purposefully expensive function?

besides, though password is never stored as text in db, one could argue it's bad form to let server touch the raw password at all regardless of TLS.

**Reply by ganigeorgiev** [reply]
>  so that you don't need server resource to run a purposefully expensive function?

You still have to run it on verify.

> besides, though password is never stored as text in db, one could argue it's bad form to let server touch the raw password at all regardless of TLS.

I don't want to argue. The benefits of having a custom bcrypt implementation are negligible and it is very error prone.

**Reply by MEF-01** [reply]
Most kind - is there an actual example/template .CSV file whose structure/content we can emulate exactly - the headings required in the .CSV are clear but it would be good to see real example of the actual format of the record entries (eg need to be in quotations or literal etc ??)

**Reply by ganigeorgiev** [reply]
@MEF-01 

More details about the `sqlite3 .import` command you can find in https://sqlite.com/cli.html#importing_files_as_csv_or_other_formats.


Since there is no information about your columns and in what order are, here is an example using a temp table so that you can insert the records in whatever columns order you have:

1. Create a CSV similar to:
    ```
    id,username,email,emailVisibility,verified,passwordHash,tokenKey,created,updated
    u1,user1,test1@example.com,false,true,$2a$12$txhm58Sty7fGgOqY4jWPHumn0I1Rq33M0II/uWK3xXl37mdXz11LK,mn0I1Rq33M0IIuWKtxhm58Sty7fGgOjW,2023-03-21 23:00:00.123Z,2023-03-21 23:00:00.123Z
    u2,user2,test2@example.com,false,false,$2a$12$txhm58Sty7fGgOqY4jWPHumn0I1Rq33M0II/uWK3xXl37mdXz11LK,n0I1Rq3y7fGgtmOjW3M0IIuWKxhm58St,2023-03-21 23:00:00.123Z,2023-03-21 23:00:00.123Z
    ```

2. Run `sqlite3 pb_data/data.db`

3. Run `.import --csv users.csv _users` (this will automatically create a temp `_users` table with the first row as columns)

4. Transfer the data to the original `users` table:
    ```sql
    INSERT INTO users (id, username, email, emailVisibility, verified, passwordHash, tokenKey, created, updated)
    SELECT * FROM _users;

    DROP TABLE _users;
    ```

**Reply by nuts-n-bits** [reply]
> You still have to run it on verify.

I see that I didn't make my point clear. I'm saying we could do this:

```
// client
let derived_key = bcrypt(raw_passwd)
authenticate("https://server.com/auth", username, derived_key)

// server
let (username, derived_key) = // from HTTP request
fetch_from_db(username, derived_key)
```
i.e. the server has no knowledge of the password hash function, it just accepts its output without further processing and match it directly with the auth database. No expensive function on server, no touching raw password, no custom bcrypt implementation. 

Only caveat being that since now bcrypt is client only, you have to maybe apply one additional salt on the server, with HMAC:

```
// server
let (username, derived_key) = // from HTTP request
let derived_key_salted = hmac_sha256(msg = derived_key, key = salt)
fetch_from_db(username, derived_key_salted)
```
And HMAC is vastly faster than bcrypt. Memory hard key derivation imho is just not server's burden to carry. Performance aside it could be a DoS attack surface too.

**Reply by ganigeorgiev** [reply]
For me this approach complicate a lot of things without bringing any benefit for the end user.
Not to mention that it also lower the security in case your server get hacked because practically the hash will end up like a "plain password" since the attacker will have access to the hmac key.

>  Performance aside it could be a DoS attack surface too.

You can dos the server by just sending requests to the health endpoint. If you want dos protection you need other mechanisms anyway (ideally on a network level before even reaching the app server).

**Reply by nuts-n-bits** [reply]
> For me this approach complicate a lot of things without bringing any benefit for the end user. 

You are downgrading bcrypt cost parameter due to performance. With the load shifted to client, you can keep the cost parameter. That's a benefit. 

> Not to mention that it also lowers the security ...

<s>Yes the naive proposal does have this problem, but all you need is another layer of sha2 to prevent it.</s>
I got carried away. In fact you don't get to reverse the HMAC even if you can see the salt, so this is a non-problem.

**Reply by ganigeorgiev** [reply]
I don't see how it won't lower the security since you'll have to store the HMAC key somewhere.
Not to mention that if we perform 1:1 string comparision against the database we are also risking timing attacks.

In any case, I don't want to argue, nor have any plans to hash the passwords client-side.

If you think that the password handling could be improved, feel free to file a  new discussion with more details, since this started to become too off-topic to the question and unnecessary spam the author.

**Reply by nuts-n-bits** [reply]
> If you think that the password handling could be improved, feel free to file a new discussion with more details, since this started to become too off-topic to the question and unnecessary spam the author.

Got it
